server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/reactivestreams/FileRangePublisher.scala

```diff
@@ -0,0 +1,84 @@
+package sttp.tapir.server.netty.internal.reactivestreams
+
+import io.netty.buffer.Unpooled
+import io.netty.handler.codec.http.{DefaultHttpContent, HttpContent}
+import org.reactivestreams.{Publisher, Subscriber, Subscription}
+import sttp.tapir.FileRange
+
+import java.nio.ByteBuffer
+import java.nio.channels.{AsynchronousFileChannel, CompletionHandler}
+import java.nio.file.StandardOpenOption
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
+
+class FileRangePublisher(fileRange: FileRange, chunkSize: Int) extends Publisher[HttpContent] {
+  override def subscribe(subscriber: Subscriber[_ >: HttpContent]): Unit = {
+    if (subscriber == null) throw new NullPointerException("Subscriber cannot be null")
+    val subscription = new FileRangeSubscription(subscriber, fileRange, chunkSize)
+    subscriber.onSubscribe(subscription)
+  }
+
+  private class FileRangeSubscription(subscriber: Subscriber[_ >: HttpContent], fileRange: FileRange, chunkSize: Int) extends Subscription {
+    private lazy val channel: AsynchronousFileChannel = AsynchronousFileChannel.open(fileRange.file.toPath(), StandardOpenOption.READ)
+    private val demand = new AtomicLong(0L)
+    private val position = new AtomicLong(fileRange.range.flatMap(_.start).getOrElse(0L))
+    private val buffer: ByteBuffer = ByteBuffer.allocate(chunkSize)
+    private val isCompleted = new AtomicBoolean(false)
+    private val readingInProgress = new AtomicBoolean(false)
+
+    override def request(n: Long): Unit = {
+      if (n <= 0) subscriber.onError(new IllegalArgumentException("§3.9: n must be greater than 0"))
+      else {
+        demand.addAndGet(n)
+        readNextChunkIfNeeded()
+      }
+    }
+
+    private def readNextChunkIfNeeded(): Unit = {
+      if (demand.get() > 0 && !isCompleted.get() && readingInProgress.compareAndSet(false, true)) {
+        val pos = position.get()
+        val expectedBytes: Int = fileRange.range.flatMap(_.end) match {
+          case Some(endPos) if pos + chunkSize > endPos => (endPos - pos + 1).toInt
+          case _                                        => chunkSize
+        }
+        buffer.clear()
+        channel.read(
+          buffer,
+          pos,
+          null,
+          new CompletionHandler[Integer, Void] {
+            override def completed(bytesRead: Integer, attachment: Void): Unit = {
+              if (bytesRead == -1) {
+                cancel()
+                subscriber.onComplete()
+              } else {
+                val bytesToRead = Math.min(bytesRead, expectedBytes)
+                buffer.flip()
```

Comment:
Although we manipulate `buffer` directly from another thread (callback to `read`), we are guaranteed that this is the only thread that operates on it as long as `readingInProgress` is `true`.

Reply:
Good point, visibility may not work well. Maybe it's better to allocate a new buffer each time.

Reply:
We are sending a newly allocated byte array, not the ByteBuf:
```scala
val bytes = new Array[Byte](bytesToRead)
buffer.get(bytes)
position.addAndGet(bytesToRead.toLong)
subscriber.onNext(new DefaultHttpContent(Unpooled.wrappedBuffer(bytes)))
```
but anyway I'm concerned if it's safe to reuse the buffer. I'll allocate a new one each time, and possibly skip the additional array.

Reply:
On the other hand - the very nature of `channel.read` requires us to pass a buffer, which we'll handle in the `completed` callback on another thread. It's unavoidable to share it between threads, right? The only difference is that in the current solution we are doing it multiple times between multiple threads.

Reply:
Ah yes I see, so the buffer will only ever be accessed by one thread, as this is guarded by our CAS logic. And you are right, we're passing the buffer to `channel.read`, so it must make sure that the contents are available when calling the callback. So all is fine I guess :)

I'd just add a lot more comments explaining how various data shared between threads is protected, and why things indeed work.